libs/features/mas/commerce/src/mas-element.js

```diff
@@ -0,0 +1,279 @@
+import {
+    CLASS_NAME_FAILED,
+    CLASS_NAME_PENDING,
+    CLASS_NAME_RESOLVED,
+    EVENT_TYPE_FAILED,
+    EVENT_TYPE_PENDING,
+    EVENT_TYPE_RESOLVED,
+    STATE_FAILED,
+    STATE_PENDING,
+    STATE_RESOLVED,
+} from './constants.js';
+import { ignore } from './external.js';
+import { discoverService, setImmediate, useService } from './utilities.js';
+
+const MasElementConstants = {
+    CLASS_NAME_FAILED,
+    CLASS_NAME_PENDING,
+    CLASS_NAME_RESOLVED,
+    EVENT_TYPE_FAILED,
+    EVENT_TYPE_PENDING,
+    EVENT_TYPE_RESOLVED,
+    STATE_FAILED,
+    STATE_PENDING,
+    STATE_RESOLVED,
+};
+
+const StateClassName = {
+    [STATE_FAILED]: CLASS_NAME_FAILED,
+    [STATE_PENDING]: CLASS_NAME_PENDING,
+    [STATE_RESOLVED]: CLASS_NAME_RESOLVED,
+};
+
+const StateEventType = {
+    [STATE_FAILED]: EVENT_TYPE_FAILED,
+    [STATE_PENDING]: EVENT_TYPE_PENDING,
+    [STATE_RESOLVED]: EVENT_TYPE_RESOLVED,
+};
+
+export class MasElement {
+    changes = new Map();
+    connected = false;
+    dispose = ignore;
+    error = undefined;
+    log = undefined;
+    options = undefined;
+    promises = [];
+    state =STATE_PENDING;
+    timer = null;
+    value = undefined;
+    version = 0;
+
+    wrapperElement;
+    
+    constructor(wrapperElement) {
+      this.wrapperElement = wrapperElement;
+    }
+
+    update() {      
+      [STATE_FAILED, STATE_PENDING, STATE_RESOLVED].forEach((state) => {
```

Comment:
IMO `STATE_PENDING` we should plan to remove, as it causes unnecessary updates.
the absence of resolved/rejected is by default pending state.

Reply:
we can do it another time.